### PR TITLE
WIP: integrated booking by API

### PR DESCRIPTION
**Pull request by pierrecamilleri**
This PR is the rebased version of #6.

Add specs for booking by API without switching platforms : POST/PATCH/GET booking.

Created another branch as I was afraid force pushing may lose some comments. 

# Edits (2nd attempt)

- Remove PUT /bookings/{BookingId} endpoint
- Transform Booking ID into UUID
- passengerPickup[Lat|Lng|Date] and passengerDrop[Lat|Lng] are mandatory in a Booking (cf [this comment](https://github.com/fabmob/standard-covoiturage/pull/6#discussion_r863604624))
- Changing the POST \messages endpoint slightly : 
  - `to` and `from` attributes with `UserId` objects.
  - `UserId` is a subset of `User` object (with `id`, `alias`, and added an `operator` field such as in #17)
  - Optional `journeyId` and `bookingId` reference with the message
- Optional `journeyId` on the Booking object.
- Added an optional `message` attribute to the PATCH endpoint, to be able to explain the status change, especially for cancellation.

<details>
  <summary>[Hidden] Details of the first try</summary>

About [my additions](https://github.com/fabmob/standard-covoiturage/pull/30/files/c353bc1661fb9845a81a5f78eb3dacf651fa3002..659753b52f174ecd5285d711dbd4476d1f8aa56f) :
- Remove PUT /bookings/{BookingId} endpoint
- Transform Booking ID into UUID
- Instead of repeating fields, I integrated `PassengerTrip` or `DriverTrip` + `JourneySchedule`. It makes sense to me to repeat this journey information, as the objective is to book a *journey*, and I was a bit bothered not to have e.g. driver information. In the spec, I plan to mention that validating a booking request is about agreeing on its required fields. As a follow-up : 
    - I put the journey `webUrl` outside of the `Trip` object as it should not be part of trip information (and had no sense in booking context)
    - I removed `passengerId` and `journeyId` : if all the information is repeated, having an ID is actually useless.
    - I removed the operator / maas urls, which where only needed to make the (ID, url) pair unique (from my understanding).
- Added a userIdToken for authentification, implying the use of MaaS Connect even for API booking.
- Added an optional `message` attribute to the PATCH endpoint, to be able to explain the status change, especially for cancellation. See below however, its relevance depends on if PATCH endpoint is used bidirectionnaly or not...
</details>

# Related questions

1. Information flow after the booking 

It is not clear to me if we would need asynchronous information push from the carsharing operator to the MaaS platform, but I think at least for status changes (e.g. cancellation !). I would suggest to have POST, GET /bookings endpoints only implemented by the carsharing operator, and I see two options : 
- either PATCH is implemented by both the MaaS platform and the carsharing operator.
- either PATCH is implemented by the carsharing operator and the latter uses the booking feed POST /booking_events like for booking by deep link.

<details>
  <summary>2. Agreement on price [RESOLVED]</summary>

(Initial comment [here](https://github.com/fabmob/standard-covoiturage/pull/6#discussion_r874568508)) 
In the search, the carsharing operator gives a price *estimate* (without knowing who is the passenger), and there are no more exchanges between carsharing operator and MaaS before the MaaS has to create a `Booking` object with definitive price. Will the MaaS always be able to determine the right price by itself with only this estimate ?
</details>

3. Exchange of user information

* I would suggest to agree on the minimal requirements to identify a user (e.g. `id` + `operator`) and use this everywhere, in conjunction with a GET /user endpoint. What do you think ?
* It also seems that the deeplink booking object is missing driver and passenger information.

4. `distance` attribute imutability

The definition of distance seems not consistent with the immutability of the booking object that has been decided. Their is currently no way to update this attribute with the "actual distance travelled" after the trip.
> Carpooling distance in meters. When the booking is COMPLETED or VALIDATED, this is the actual distance travelled if available.

5. Use POST /messages for deeplink ?

Any interest in sharing this endpoint for deeplink and API booking ?

### CODE REVIEW

File: standard-covoiturage_openapi.yaml [hunk 1]
```diff
@@ -176,6 +176,171 @@ paths:
...
+          $ref: '#/components/responses/BadRequest'
+        401:
+          $ref: '#/components/responses/Unauthorized'
+        429:
+          $ref: '#/components/responses/TooManyRequests'
+        500:
+          $ref: '#/components/responses/InternalServerError'
+
+  /bookings:
+    post:
```
Comment (osarrat) [inline standard-covoiturage_openapi.yaml:219]: **Review on the POST /booking endpoint :** 
- The `message` parameter should not be compulsory.
- Why having added a `type` parameter to the booking ? It seems irrelevant to me.
- Why having added a `userIdToken` for MaaS Connect since the core objective of the booking by API is to book a journey _without_ having to connect to another service ?

**Review by osarrat — CHANGES_REQUESTED** [review]
Thanks @pierrecamilleri for this attempt to finish this work !

I have several request for changes, and I remain open for discussion about them.

File: standard-covoiturage_openapi.yaml [hunk 2]
```diff
@@ -176,6 +176,171 @@ paths:
...
+        required: true
+        schema:
+          $ref: '#/components/schemas/bookingId'
+      - name: status
+        description: New status of the Booking.
+        in: query
+        required: true
+        schema:
+          $ref: '#/components/schemas/bookingStatus'
+      - message:
```
[the remark this answers is not in the document]
Comment (pierrecamilleri) [reply standard-covoiturage_openapi.yaml:266]: Oopsie you are right, I pushed unfinished work in my last commit. Sorry for that, I make the correction.

File: standard-covoiturage_openapi.yaml (hunk 1 above)
Comment (pierrecamilleri) [reply standard-covoiturage_openapi.yaml:219]: > The message parameter should not be compulsory.
Ok

> type parameter to the booking ? It seems irrelevant to me.
It came as a byproduct of puting a whole `JourneySchedule` into the booking. I'll think about how to reorganize things.

> Why having added a userIdToken for MaaS Connect since the core objective of the booking by API is to book a journey without having to connect to another service ?

Ok, with this comment + the comment on the price, I think I get the idea. In the booking by API, the carsharing operator doesn't have to know the details about the passenger (so, no specific authentification).

**Comment by pierrecamilleri** [comment]
Thank you for your review !

>  the the use case of interoperability between TWO carpooling
> API service must be designed symmetrically

My understanding was of a weaker objective. From my point of view, I understood that it MAY be IMPLEMENTED symmetrically, but I don't think we need a symmetric design here. (However, symmetric PATCH seems fine to me).

>  the price returned by a GET /journeys is NOT an estimate. It is actually the price expected by the owner of the journey.

Ok, it was still not clear to me. I was somehow thinking the carsharing operator was fixing the price. So the price paid by the platform booking the journey to the carsharing operator does actually NOT depend on the passenger in any way. Is that right ?

File: standard-covoiturage_openapi.yaml [hunk 3]
```diff
@@ -176,6 +176,171 @@ paths:
         500:
           $ref: '#/components/responses/InternalServerError'
 
+  /messages:
+    post:
```
[the remark this answers is not in the document]
Comment (pierrecamilleri) [reply standard-covoiturage_openapi.yaml:180]: hum, yes, my mistake, removed stuff from `ConnectionMessage` with the `\bookings` endpoint in mind.

File: standard-covoiturage_openapi.yaml [hunk 4]
```diff
@@ -756,6 +951,84 @@ components:
...
+          type: string
+          description: URL of the booking on the webservice provider platform.
+        price:
+          $ref: '#/components/schemas/Price'
+        car:
+          $ref: '#/components/schemas/Car'
+        message:
+          type: string
+          maxLength: 500
+          description: Free text content of a message. The message can contain all the details (phone number, email, etc.) allowing the recipient to call back the sender in order to carpool with him/her.
```
Comment (osarrat) [inline standard-covoiturage_openapi.yaml:1014]: Since `Booking` is no more inherited from `ConnectionMessage`, let's change the description of the `Booking.message` attribute to simply :  

> Free text content of a message.

The ~The message can contain all the details (phone number, email, etc.) allowing the recipient to call back the sender in order to carpool with him/her.~ part car be removed because irrelevant in case of a Booking.

**Review by osarrat — CHANGES_REQUESTED** [review]
We're close to it!

**Comment by osarrat** [comment]
Regarding your remaining first blind spot  :  

> 1.  Information flow after the booking
> 
> It is not clear to me if we would need asynchronous information push from the carsharing operator to the MaaS platform, but I think at least for status changes (e.g. cancellation !). I would suggest to have POST, GET /bookings endpoints only implemented by the carsharing operator, and I see two options :
> 
> *   either PATCH is implemented by both the MaaS platform and the carsharing operator.
> *   either PATCH is implemented by the carsharing operator and the latter uses the booking feed POST /booking\_events like for booking by deep link.

The POST /booking\_events cannot be used in case of a booking by API. They don't reply to the same use case. For example, in case of a MaaS Platform implementing the Booking by API use case, it MUST implement a solution for message exchange between its user (usually passenger) in its MaaS Platform and the user (usually driver) in the carpooling platform. Exchanging message in the PATCH /bookings is relevant for this use case. But implementing the required message exchange in the POST /booking\_events is irrelevant. Thus, we cannot use POST /booking\_events anytime in the Booking by API use case.

File: standard-covoiturage_openapi.yaml [hunk 5]
```diff
@@ -756,6 +951,84 @@ components:
           type: string
           description: Brand of the car.
 
+
+    Booking:
```
[the remark this answers is not in the document]
Comment (pierrecamilleri) [reply standard-covoiturage_openapi.yaml:953]: The `driverJourneyId`'s and `passengerJourneyId`'s operators should be available through the `driver.operator` and `passenger.operators` properties. Is it ok like that ?

**Comment by osarrat** [comment]
Thanks for this last work @pierrecamilleri , and it was ok for me to merge it.  
I made a secondary review directly here : https://github.com/fabmob/standard-covoiturage/pull/2#pullrequestreview-1005796179
